### PR TITLE
Proposal to change PR#9 to trivago tgo

### PR DESCRIPTION
**Pull request by arnecls**
## Reasons for the change:

TL;DR: Make the new regex functionality match the behaviour of current parsing modes.

The way the regex function has been build makes it behave differently than the other parsing modes.
The main problem - as already stated in the PR - is the way that incomplete data is (not) handled. This is required as data streams might not return all data with a single read and might thus require multiple read passes to generate one complete message.
In addition to this It should be up to the caller when and how to deal with incomplete data after multiple reads, e.g. by utilising a timeout. Existing parsing modes already provide such a behaviour, at least when using the ReadOne function.

Another change is the way delimiters are detected. Delimiters before your PR used "end of line markers".  Your PR introduced "start of line markers" which might not always be the desired usecase. As of this there are now two modes for start and end of line search.

Another benefit of the changes in this PR is, that the parsing algorithm is now a lot simpler, leaving less room for bugs.

## Why this PR here and not in trivago/tgo

This is meant to be a suggestion for improvements of [PR 9 in trivago/tgo](https://github.com/trivago/tgo/pull/9).

## Changes:

- Rename multiline to regex.
- Split into two modes: find start, find end.
- Leave unmatched data in the buffer to match behaviour of previous parsing modes.
- Add a possibility to get incomplete data from the buffer.
- Add edge cases to unit tests

### CODE REVIEW

**Comment by serii833** [comment]
What is the best way to handle this PR? 
Should I merge it to my fork and update PR#9 to trivago/tgo or make a new PR to trivago/tgo?

**Comment by arnecls** [comment]
The reasons I did the PR here are
- we can discuss the changes here (and I don't simply overwrite your stuff)
- you have the chance to do adjustments after merging this
- the PR is still attributed to you, as you did the initial work

So if you're fine - merge it and the PR#9 should automatically update.
